GVFS/GVFS.UnitTests/Tracing/QueuedPipeStringWriterTests.cs

```diff
@@ -0,0 +1,164 @@
+                () => null); // Not important since the test will never enqueue any messaged to write
+
+            // Capture event invocations
+            List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls = new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+            writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+            // Try to write some dummy data
+            writer.Start();
+            writer.Stop();
+
+            Assert.AreEqual(1, eventCalls.Count);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[0].State);
+        }
+
+        [TestCase]
+        public void MissingPipe_RaisesStateFailing()
+        {
+            const string inputMessage = "FooBar";
+            QueuedPipeStringWriter writer = new QueuedPipeStringWriter(() => throw new Exception("Failing pipe connection"));
+
+            // Capture event invocations
+            List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls = new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+            writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+            // Try to write some dummy data
+            writer.Start();
+            writer.Enqueue(inputMessage);
+            writer.Stop();
+
+            Assert.AreEqual(2, eventCalls.Count);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Failing, eventCalls[0].State);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[1].State);
+        }
+
+        [TestCase]
+        public void GoodPipe_WritesDataAndRaisesStateHealthy()
+        {
+            const string inputMessage = "FooBar";
+            byte[] expectedData = { 0x46, 0x6F, 0x6F, 0x42, 0x61, 0x72, (byte)'\n' };
+
+            string pipeName = Guid.NewGuid().ToString("N");
+
+            using (TestPipeWorker pipeWorker = new TestPipeWorker(pipeName, PipeTransmissionMode.Byte))
+            {
+                QueuedPipeStringWriter writer = new QueuedPipeStringWriter(() =>
+                    new NamedPipeClientStream(".", pipeName, PipeDirection.Out));
+
+                // Capture event invocations
+                List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls =
+                    new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+                writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+                // Try to write some dummy data
+                writer.Start();
+                writer.Enqueue(inputMessage);
+                writer.Stop();
+
+                byte[] actualData = pipeWorker.GetAllData();
+                CollectionAssert.AreEqual(expectedData, actualData);
+
+                Assert.AreEqual(2, eventCalls.Count);
+                Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Healthy, eventCalls[0].State);
```

Comment:
There is logic already in JsonTracer to only fire 'recovery' and 'failure' once per change.
I could move this logic, or rather duplicate it here in the `QueuedPipeStringWriter` too, so that state change events only fire once per change.
The logic in the `JsonTracer` would need to remain to stop reporting repeated recovery/failure for other tracers.